src/Raven.Client/Documents/Session/SessionCountersBase.cs

```diff
@@ -13,16 +13,31 @@ namespace Raven.Client.Documents.Session
     /// <summary>
     /// Abstract implementation for in memory session operations
     /// </summary>
-    public abstract class DocumentSessionCountersBase : AdvancedSessionExtentionBase
+    public abstract class SessionCountersBase 
     {
-        protected DocumentSessionCountersBase(InMemoryDocumentSessionOperations session) : base(session)
-        {
-        }
+        protected string DocId;
+        protected InMemoryDocumentSessionOperations Session;
 
-        public void Increment(string documentId, string counter, long delta = 1)
+        protected SessionCountersBase(InMemoryDocumentSessionOperations session, string documentId)
         {
+            Session = session;
+
             if (string.IsNullOrWhiteSpace(documentId))
                 throw new ArgumentNullException(nameof(documentId));
+            DocId = documentId;
+        }
+
+        protected SessionCountersBase(InMemoryDocumentSessionOperations session, object entity)
+        {
+            Session = session;
+
+            if (Session.DocumentsByEntity.TryGetValue(entity, out DocumentInfo document) == false)
```

Comment:
Okay, so this is resolved once and we aren't holding on to it. Make sense.
However, I wonder if we can take advantage of this to skip the `GotAll` note. If we know that we have the metadata and are missing counters, we can check with the metadata in the session to know if we should get it.

Reply:
done